lib/compress/zstd_lazy.c

```diff
@@ -531,7 +555,47 @@ size_t ZSTD_HcFindBestMatch_generic (
         matchIndex = NEXT_IN_CHAIN(matchIndex, chainMask);
     }
 
-    if (dictMode == ZSTD_dictMatchState) {
+    if (dictMode == ZSTD_dictMatchState && ms->dictMatchState->enableDedicatedDictSearch) {
```

Comment:
Does that sound possible to extract the dictionary search routine into its own function ?
This would centralize its implementation.